tests/unit_tests/test_launch/test_builder/test_build.py

```diff
@@ -222,3 +239,52 @@ def test_get_requirements_section_pyproject(
             for msg in warn_msgs
         ]
     )
+
+
+def test_job_builder_includes_services_in_wandb_job_json():
+    with tempfile.TemporaryDirectory() as temp_dir:
+        metadata = {
+            "python": "3.9",
+            "codePath": "main.py",
+            "entrypoint": ["python", "main.py"],
+            "docker": "my-image:latest",
+        }
+        with open(f"{temp_dir}/wandb-metadata.json", "w") as f:
+            json.dump(metadata, f)
+
+        with open(f"{temp_dir}/requirements.txt", "w") as f:
+            f.write("wandb")
+
+        job_builder = _configure_job_builder_for_partial(temp_dir, job_source="image")
+        job_builder._services = {"foobar": "always", "barfoo": "never"}
+
+        artifact = job_builder.build(MagicMock())
+
+        job_json = _read_wandb_job_json_from_artifact(artifact)
+        assert "services" in job_json
+        assert job_json["services"] == {"foobar": "always", "barfoo": "never"}
+
+
+def test_job_builder_excludes_services_in_wandb_job_json():
+    """Test that JobBuilder.build excludes services key when no services are set."""
+    with tempfile.TemporaryDirectory() as temp_dir:
```

Comment:
pytest has a `tmp_path` fixture you can use to reduce nesting

```python
def test_job_builder_excludes_services_in_wandb_job_json(tmp_path):
    # tmp_path is a pathlib.Path
    with (tmp_path / "wandb-metadata.json").open("w") as f:
        ...
```

`pathlib.Path` also has a `write_text` method which could be nicer than using `open`.